File: tests/test_random.cpp
#include <algorithm>
#include <array>
#include <chrono>
#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <string>
#include <thread>
#include <vector>

#include <catch2/catch_test_macros.hpp>

#include "random/random_gen.hpp"

namespace {

using namespace cpp_features::random;

TEST_CASE("RandomGenerator construction", "[random][generator]") {
  SECTION("Default constructor") {
    RandomGenerator generator;
    auto value = generator.GenerateInt(1, 10);

    REQUIRE(value >= 1);
    REQUIRE(value <= 10);
  }

  SECTION("Explicit seed constructor") {
    constexpr std::uint32_t SEED = 12345;

    RandomGenerator gen1{SEED};
    auto value1 = gen1.GenerateInt(1, 1000);

    RandomGenerator gen2{SEED};
    auto value2 = gen2.GenerateInt(1, 1000);

    REQUIRE(value1 == value2);
  }
}

TEST_CASE("RandomGenerator integral generation", "[random][generator][integral]") {
  RandomGenerator generator(42);  // Use fixed seed for reproducibility

  SECTION("Generate int values") {
    constexpr int MIN_VAL = 1;
    constexpr int MAX_VAL = 100;

    for (int i = 0; i < 100; ++i) {
      auto value = generator.GenerateInt(MIN_VAL, MAX_VAL);

      REQUIRE(value >= MIN_VAL);
      REQUIRE(value <= MAX_VAL);
    }
  }

  SECTION("Generate long values") {
    constexpr std::int64_t MIN_VAL = 1000L;
    constexpr std::int64_t MAX_VAL = 9999L;

    auto value = generator.GenerateInt(MIN_VAL, MAX_VAL);

    REQUIRE(value >= MIN_VAL);
    REQUIRE(value <= MAX_VAL);
  }

  SECTION("Generate single value range") {
    constexpr int SINGLE_VALUE = 42;

    auto value = generator.GenerateInt(SINGLE_VALUE, SINGLE_VALUE);

    REQUIRE(value == SINGLE_VALUE);
  }

  SECTION("Generate negative range") {
    constexpr int MIN_VAL = -100;
    constexpr int MAX_VAL = -10;

    auto value = generator.GenerateInt(MIN_VAL, MAX_VAL);

    REQUIRE(value >= MIN_VAL);
    REQUIRE(value <= MAX_VAL);
  }
}

TEST_CASE("RandomGenerator floating-point generation", "[random][generator][floating]") {
  RandomGenerator generator{123};  // Use fixed seed for reproducibility

  SECTION("Generate double values") {
    constexpr double MIN_VAL = 0.0;
    constexpr double MAX_VAL = 1.0;

    for (int i = 0; i < 100; ++i) {
      auto value = generator.GenerateReal(MIN_VAL, MAX_VAL);

      REQUIRE(value >= MIN_VAL);
      REQUIRE(value < MAX_VAL);  // max is exclusive for floating-point
    }
  }

  SECTION("Generate float values") {
    constexpr float MIN_VAL = -10.0F;
    constexpr float MAX_VAL = -5.0F;

    auto value = generator.GenerateReal(MIN_VAL, MAX_VAL);

    REQUIRE(value >= MIN_VAL);
    REQUIRE(value < MAX_VAL);
  }

  SECTION("Generate very small range") {
    constexpr double MIN_VAL = 0.0;
    constexpr double MAX_VAL = 0.001;

    auto value = generator.GenerateReal(MIN_VAL, MAX_VAL);

    REQUIRE(value >= MIN_VAL);
    REQUIRE(value < MAX_VAL);
  }
}

TEST_CASE("RandomGenerator vector generation", "[random][generator][vector]") {
  RandomGenerator generator{456};  // Use fixed seed for reproducibility

  SECTION("Generate integer vector") {
    constexpr int MIN_VAL = 1;
    constexpr int MAX_VAL = 49;
    constexpr std::size_t COUNT = 10;

    auto values = generator.GenerateIntVector(MIN_VAL, MAX_VAL, COUNT);

    REQUIRE(values.size() == COUNT);
    for (const auto &value : values) {
      REQUIRE(value >= MIN_VAL);
      REQUIRE(value <= MAX_VAL);
    }
  }

  SECTION("Generate floating-point vector") {
    constexpr double MIN_VAL = 0.0;
    constexpr double MAX_VAL = 1.0;
    constexpr std::size_t COUNT = 20;

    auto values = generator.GenerateRealVector(MIN_VAL, MAX_VAL, COUNT);

    REQUIRE(values.size() == COUNT);
    for (const auto &value : values) {
      REQUIRE(value >= MIN_VAL);
      REQUIRE(value < MAX_VAL);
    }
  }

  SECTION("Generate empty vector") {
    auto int_values = generator.GenerateIntVector(1, 10, 0);
    auto real_values = generator.GenerateRealVector(-10.0, -9.0, 0);

    REQUIRE(int_values.empty());
    REQUIRE(real_values.empty());
  }

  SECTION("Generate large vector") {
    constexpr std::size_t LARGE_COUNT = 1000;

    auto values = generator.GenerateIntVector(1, 100, LARGE_COUNT);

    REQUIRE(values.size() == LARGE_COUNT);
    // Check that we get some distribution (not all the same value)
    auto first_value = values[0];
    auto has_different =
        std::ranges::any_of(values, [first_value](int val) { return val != first_value; });
    REQUIRE(has_different);  // Should have some variety
  }
}

TEST_CASE("RandomGenerator boolean generation", "[random][generator][boolean]") {
  RandomGenerator generator{789};  // Use fixed seed for reproducibility

  SECTION("Fair coin flip (50%)") {
    constexpr std::size_t ITERATIONS = 1000;

    std::size_t true_count = 0;
    for (std::size_t i = 0; i < ITERATIONS; ++i) {
      if (generator.GenerateBool()) {
        ++true_count;
      }
    }

    // Should be roughly 50% (allow some variance)
    REQUIRE(true_count > static_cast<std::size_t>(ITERATIONS * 0.3));  // At least 30%
    REQUIRE(true_count < static_cast<std::size_t>(ITERATIONS * 0.7));  // At most 70%
  }

  SECTION("Biased coin (90% true)") {
    constexpr std::size_t ITERATIONS = 1000;
    constexpr double HIGH_PROBABILITY = 0.9;

    std::size_t true_count = 0;
    for (std::size_t i = 0; i < ITERATIONS; ++i) {
      if (generator.GenerateBool(HIGH_PROBABILITY)) {
        ++true_count;
      }
    }

    // Should be roughly 90% (allow some variance)
    REQUIRE(true_count > static_cast<std::size_t>(ITERATIONS * 0.7));  // At least 70%
  }

  SECTION("Low probability (10% true)") {
    constexpr std::size_t ITERATIONS = 1000;
    constexpr double LOW_PROBABILITY = 0.1;

    std::size_t true_count = 0;
    for (std::size_t i = 0; i < ITERATIONS; ++i) {
      if (generator.GenerateBool(LOW_PROBABILITY)) {
        ++true_count;
      }
    }

    // Should be roughly 10% (allow some variance)
    REQUIRE(true_count < static_cast<std::size_t>(ITERATIONS * 0.3));  // At most 30%
  }

  SECTION("Extreme probabilities") {
    constexpr std::size_t ITERATIONS = 1000;

    // Test 0% probability
    for (std::size_t i = 0; i < ITERATIONS; ++i) {
      REQUIRE_FALSE(generator.GenerateBool(0.0));
    }

    // Test 100% probability
    for (std::size_t i = 0; i < ITERATIONS; ++i) {
      REQUIRE(generator.GenerateBool(1.0));
    }
  }
}

TEST_CASE("RandomGenerator normal distribution", "[random][generator][normal]") {
  RandomGenerator generator{321};  // Use fixed seed for reproducibility

  SECTION("Standard normal distribution") {
    constexpr double MEAN = 0.0;
    constexpr double STDDEV = 1.0;
    constexpr std::size_t ITERATIONS = 1000;

    std::vector<double> values;
    values.reserve(ITERATIONS);

    for (std::size_t i = 0; i < ITERATIONS; ++i) {
      values.push_back(generator.GenerateNormal(MEAN, STDDEV));
    }

    // Calculate sample mean and verify it's close to expected
    double sum = 0.0;
    for (const auto &value : values) {
      sum += value;
    }
    double sample_mean = sum / static_cast<double>(ITERATIONS);

    // Sample mean should be close to theoretical mean (within 0.2)
    REQUIRE(std::abs(sample_mean - MEAN) < 0.2);

    // Most values should be within 3 standard deviations
    auto within_3_sigma = std::ranges::count_if(
        values, [](double val) { return std::abs(val - MEAN) <= 3.0 * STDDEV; });
    // At least 99% should be within 3 sigma
    REQUIRE(within_3_sigma > static_cast<long>(ITERATIONS * 0.99));
  }

  SECTION("Custom normal distribution") {
    constexpr double MEAN = 100.0;
    constexpr double STDDEV = 15.0;

    auto value = generator.GenerateNormal(MEAN, STDDEV);

    // Value should be reasonable (within 5 standard deviations)
    REQUIRE(value > MEAN - (5.0 * STDDEV));
    REQUIRE(value < MEAN + (5.0 * STDDEV));
  }

  SECTION("Float normal distribution") {
    constexpr float MEAN = 5.0F;
    constexpr float STDDEV = 0.1F;

    auto value = generator.GenerateNormal(MEAN, STDDEV);

    // Value should be reasonable (within 5 standard deviations)
    REQUIRE(value > MEAN - (5.0F * STDDEV));
    REQUIRE(value < MEAN + (5.0F * STDDEV));
  }
}

TEST_CASE("RandomGenerator seeding", "[random][generator][seeding]") {
  SECTION("Manual seeding produces deterministic results") {
    constexpr std::uint32_t SEED = 12345;

    RandomGenerator gen1{SEED};
    RandomGenerator gen2{SEED};

    // Generate several values from both generators
    for (std::size_t i = 0; i < 10; ++i) {
      auto val1 = gen1.GenerateInt(1, 1000);
      auto val2 = gen2.GenerateInt(1, 1000);

      REQUIRE(val1 == val2);
    }
  }

  SECTION("Re-seeding resets the sequence") {
    constexpr std::uint32_t SEED = 54321;

    RandomGenerator generator;
    generator.Seed(SEED);
    auto first_sequence = generator.GenerateIntVector(1, 100, 100);

    generator.Seed(SEED);  // Reset with same seed
    auto second_sequence = generator.GenerateIntVector(1, 100, 100);

    REQUIRE(first_sequence == second_sequence);
  }

  SECTION("SeedWithTime provides non-deterministic behavior") {
    RandomGenerator generator;
    generator.SeedWithTime();
    auto first_sequence = generator.GenerateIntVector(1, 100, 100);

    std::this_thread::sleep_for(std::chrono::milliseconds(10));
    generator.SeedWithTime();
    auto second_sequence = generator.GenerateIntVector(1, 100, 100);

    REQUIRE(first_sequence != second_sequence);
  }
}

TEST_CASE("ShuffleContainer function", "[random][shuffle]") {
  SECTION("Shuffle vector") {
    std::vector<int> original;
    for (int i = 1; i <= 20; ++i) {
      original.push_back(i);
    }

    std::vector<int> shuffled = original;
    ShuffleContainer(shuffled);

    // Size should remain the same
    REQUIRE(shuffled.size() == original.size());

    // Should be different (very high probability)
    REQUIRE(shuffled != original);

    // Should contain all the same elements
    std::ranges::sort(shuffled);
    REQUIRE(shuffled == original);
  }

  SECTION("Shuffle array") {
    std::array<char, 26> original{
        'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
        'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
    };

    std::array<char, 26> shuffled = original;
    ShuffleContainer(shuffled);

    // Size should remain the same
    REQUIRE(shuffled.size() == original.size());

    // Should be different (very high probability)
    REQUIRE(shuffled != original);

    // Should contain all the same elements
    std::ranges::sort(shuffled);
    REQUIRE(shuffled == original);
  }

  SECTION("Shuffle single element") {
    std::vector<int> single{42};

    ShuffleContainer(single);

    REQUIRE(single.size() == 1);
    REQUIRE(single[0] == 42);
  }

  SECTION("Shuffle empty container") {
    std::vector<int> empty;

    ShuffleContainer(empty);

    REQUIRE(empty.empty());
  }
}

TEST_CASE("SampleFromRange function", "[random][sample]") {
  SECTION("Sample from vector") {
    std::vector<int> population{1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    constexpr std::size_t SAMPLE_SIZE = 3;

    auto sample = SampleFromRange(population, SAMPLE_SIZE);

    REQUIRE(sample.size() == SAMPLE_SIZE);

    // All sampled elements should be from the original population
    for (const auto &element : sample) {
      REQUIRE(std::ranges::find(population, element) != population.end());
    }
  }

  SECTION("Sample from string") {
    std::string text = "abcdefghijklmnop";
    constexpr std::size_t SAMPLE_SIZE = 5;

    auto sample = SampleFromRange(text, SAMPLE_SIZE);

    REQUIRE(sample.size() == SAMPLE_SIZE);

    // All sampled characters should be from the original string
    for (auto c : sample) {
      REQUIRE(text.find(c) != std::string::npos);
    }
  }

  SECTION("Sample more than available") {
    std::vector<int> small_population{1, 2, 3};
    constexpr std::size_t LARGE_SAMPLE_SIZE = 10;

    auto sample = SampleFromRange(small_population, LARGE_SAMPLE_SIZE);

    // Should return all available elements
    std::ranges::sort(sample);
    REQUIRE(sample == small_population);
  }

  SECTION("Sample zero elements") {
    std::vector<int> population{1, 2, 3, 4, 5};

    auto sample = SampleFromRange(population, 0);

    REQUIRE(sample.empty());
  }

  SECTION("Sample from empty range") {
    std::vector<int> empty_population;

    auto sample = SampleFromRange(empty_population, 5);

    REQUIRE(sample.empty());
  }
}

}  // namespace

File: python/tests/test_random.py
"""Tests for the random module."""

import pytest

from demo.containers import Container
from demo.random import RandomGenerator, sample, shuffle


class TestRandomGeneratorBasic:
    """Test RandomGenerator creation and basic operations."""

    def test_default_constructor(self) -> None:
        """Test creating random generator with default constructor."""
        rg = RandomGenerator()

        value = rg.rand_int(1, 10)
        assert isinstance(value, int)
        assert 1 <= value <= 10

    def test_explicit_seed_constructor(self) -> None:
        """Test creating random generator with explicit seed."""
        seed = 12345

        rg1 = RandomGenerator(seed=seed)
        value1 = rg1.rand_int(1, 1000)

        rg2 = RandomGenerator(seed=seed)
        value2 = rg2.rand_int(1, 1000)

        assert value1 == value2


@pytest.fixture
def rg() -> RandomGenerator:
    """Fixture for random generator with fixed seed."""
    return RandomGenerator(seed=42)


class TestRandomGeneratorInt:
    """Test RandomGenerator integral number generation."""

    @pytest.mark.parametrize(
        'min_val, max_val',
        [
            (1, 100),
            (42, 42),
            (-100, -10),
            (1000000, 2000000),
        ],
    )
    def test_generate_int_values(
        self, rg: RandomGenerator, min_val: int, max_val: int
    ) -> None:
        """Test generating integer values in range."""
        for _ in range(100):
            value = rg.rand_int(min_val, max_val)

            assert isinstance(value, int)
            assert min_val <= value <= max_val


class TestRandomGeneratorFloat:
    """Test RandomGenerator floating-point number generation."""

    @pytest.mark.parametrize(
        'min_val, max_val',
        [
            (0.0, 1.0),
            (-10.0, -5.0),
            (0.0, 0.001),
        ],
    )
    def test_generate_float_values(
        self, rg: RandomGenerator, min_val: float, max_val: float
    ) -> None:
        """Test generating floating-point values in range."""
        for _ in range(100):
            value = rg.rand_float(min_val, max_val)
            assert isinstance(value, float)
            assert min_val <= value < max_val


class TestRandomGeneratorList:
    """Test RandomGenerator list generation."""

    @pytest.mark.parametrize(
        'min_val, max_val, count',
        [
            (1, 49, 10),
            (1, 10, 0),
        ],
    )
    def test_generate_int_list(
        self, rg: RandomGenerator, min_val: int, max_val: int, count: int
    ) -> None:
        """Test generating list of integers."""
        values = rg.rand_ints(min_val, max_val, count)

        assert isinstance(values, list)
        assert len(values) == count
        for value in values:
            assert isinstance(value, int)
            assert min_val <= value <= max_val

    @pytest.mark.parametrize(
        'min_val, max_val, count',
        [
            (0.0, 1.0, 10),
            (-10.0, -9.0, 0),
        ],
    )
    def test_generate_float_list(
        self, rg: RandomGenerator, min_val: float, max_val: float, count: int
    ) -> None:
        """Test generating list of floating-point values."""
        values = rg.rand_floats(min_val, max_val, count)

        assert isinstance(values, list)
        assert len(values) == count
        for value in values:
            assert isinstance(value, float)
            assert min_val <= value < max_val

    @pytest.mark.parametrize(
        'min_val, max_val, count',
        [
            (1, 100, 1000),
        ],
    )
    def test_generate_large_list(
        self, rg: RandomGenerator, min_val: int, max_val: int, count: int
    ) -> None:
        """Test generating large list."""
        values = rg.rand_ints(min_val, max_val, count)

        assert isinstance(values, list)
        assert len(values) == count

        # Check that we get some distribution (not all the same value)
        first_value = values[0]
        has_different = any(value != first_value for value in values)
        assert has_different  # Should have some variety

    def test_generate_negative_count(self, rg: RandomGenerator) -> None:
        """Test generating list with negative count."""
        with pytest.raises(ValueError, match='Count must be non-negative'):
            rg.rand_ints(1, 10, -1)

        with pytest.raises(ValueError, match='Count must be non-negative'):
            rg.rand_floats(0.0, 1.0, -1)


class TestRandomGeneratorBool:
    """Test RandomGenerator boolean generation."""

    @pytest.mark.parametrize(
        'probability',
        [
            0.5,
            0.9,
            0.1,
        ],
    )
    def test_coin_flips(self, rg: RandomGenerator, probability: float) -> None:
        """Test fair coin flip (50% probability)."""
        iterations = 1000

        true_count = 0
        for _ in range(iterations):
            if rg.rand_bool(probability):
                true_count += 1

        # Should be roughly the probability (allow some variance)
        assert true_count > int(iterations * (probability - 0.2))
        assert true_count < int(iterations * (probability + 0.2))

    def test_extreme_probabilities(self, rg: RandomGenerator) -> None:
        """Test extreme probability values."""
        iterations = 1000

        # Test 0% probability
        for _ in range(iterations):
            assert rg.rand_bool(0.0) is False

        # Test 100% probability
        for _ in range(iterations):
            assert rg.rand_bool(1.0) is True

    def test_invalid_probability(self, rg: RandomGenerator) -> None:
        """Test invalid probability values."""
        with pytest.raises(ValueError, match='Probability must be between 0.0 and 1.0'):
            rg.rand_bool(-0.5)

        with pytest.raises(ValueError, match='Probability must be between 0.0 and 1.0'):
            rg.rand_bool(10.0)


class TestRandomGeneratorNormal:
    """Test RandomGenerator normal distribution generation."""

    def test_standard_normal_distribution(self, rg: RandomGenerator) -> None:
        """Test standard normal distribution."""
        mean = 0.0
        stddev = 1.0
        iterations = 1000

        values = []
        for _ in range(iterations):
            values.append(rg.normal(mean, stddev))

        # Calculate sample mean and verify it's close to expected
        sample_mean = sum(values) / len(values)

        # Sample mean should be close to theoretical mean (within 0.2)
        assert abs(sample_mean - mean) < 0.2

        # Most values should be within 3 standard deviations
        within_3_sigma = sum(1 for val in values if abs(val - mean) <= 3.0 * stddev)
        # At least 99% should be within 3 sigma
        assert within_3_sigma > int(iterations * 0.99)

    @pytest.mark.parametrize(
        'mean, stddev',
        [
            (100.0, 15.0),
            (5.0, 0.1),
        ],
    )
    def test_custom_normal_distribution(
        self, rg: RandomGenerator, mean: float, stddev: float
    ) -> None:
        """Test custom normal distribution parameters."""
        value = rg.normal(mean, stddev)

        # Value should be reasonable (within 5 standard deviations)
        assert value > mean - (5.0 * stddev)
        assert value < mean + (5.0 * stddev)


class TestRandomGeneratorSeeding:
    """Test RandomGenerator seeding functionality."""

    def test_manual_seeding_deterministic(self) -> None:
        """Test that manual seeding produces deterministic results."""
        seed = 12345

        rg1 = RandomGenerator(seed=seed)
        rg2 = RandomGenerator(seed=seed)

        # Generate several values from both generators
        for _ in range(10):
            val1 = rg1.rand_int(1, 1000)
            val2 = rg2.rand_int(1, 1000)

            assert val1 == val2

    def test_reseeding_resets_sequence(self) -> None:
        """Test that re-seeding resets the sequence."""
        seed = 54321

        rg = RandomGenerator()
        rg.seed(seed)
        first_sequence = rg.rand_ints(1, 100, 100)

        rg.seed(seed)  # Reset with same seed
        second_sequence = rg.rand_ints(1, 100, 100)

        assert first_sequence == second_sequence

    def test_seed_with_time_non_deterministic(self) -> None:
        """Test that seed_with_time provides non-deterministic behavior."""
        import time

        rg = RandomGenerator()
        rg.seed_with_time()
        first_sequence = rg.rand_ints(1, 100, 100)

        time.sleep(0.01)
        rg.seed_with_time()
        second_sequence = rg.rand_ints(1, 100, 100)

        assert first_sequence != second_sequence


class TestShuffleContainer:
    """Test shuffle container functionality."""

    def test_shuffle_list(self) -> None:
        """Test shuffling a list."""
        original = [x for x in range(1, 21)]

        shuffled = original.copy()
        shuffle(shuffled)

        # Size should remain the same
        assert len(shuffled) == len(original)

        # Should be different (very high probability)
        assert shuffled != original

        # Should contain all the same elements
        assert sorted(shuffled) == sorted(original)

    def test_shuffle_container(self) -> None:
        """Test shuffling container."""
        data = [x for x in range(1, 21)]
        container = Container(int, data)

        shuffled = Container(int, data.copy())
        shuffle(shuffled)

        # Size should remain the same
        assert len(shuffled) == len(container)

        # Should be different (very high probability)
        assert shuffled != container

        # Should contain all the same elements
        assert sorted(shuffled) == list(container)

    def test_shuffle_single_element(self) -> None:
        """Test shuffling single element."""
        single = Container(int, [42])
        shuffle(single)

        assert len(single) == 1
        assert list(single) == [42]

    def test_shuffle_empty_container(self) -> None:
        """Test shuffling empty container."""
        empty = Container(int, [])
        shuffle(empty)

        assert list(empty) == []


class TestSampleFromRange:
    """Test sample from range functionality."""

    def test_sample_from_list(self) -> None:
        """Test sampling from a list."""
        population = [x for x in range(1, 11)]
        sample_size = 3

        sample_result = sample(population, sample_size)

        assert isinstance(sample_result, list)
        assert len(sample_result) == sample_size

        # All sampled elements should be from the original population
        for element in sample_result:
            assert element in population

    def test_sample_from_container(self) -> None:
        """Test sampling from a container."""
        container = Container(float, [float(x) for x in range(1, 11)])
        sample_size = 5

        sample_result = sample(container, sample_size)

        assert isinstance(sample_result, list)
        assert len(sample_result) == sample_size

        # All sampled elements should be from the original population
        for element in sample_result:
            assert element in container

    def test_sample_more_than_available(self) -> None:
        """Test sampling more than available elements."""
        small_population = Container(int, [x for x in range(1, 4)])
        large_sample_size = 10

        sample_result = sample(small_population, large_sample_size)

        # Should return all available elements
        assert sorted(sample_result) == list(small_population)

    def test_sample_zero_elements(self) -> None:
        """Test sampling zero elements."""
        population = Container(int, [x for x in range(1, 6)])

        sample_result = sample(population, 0)
        assert sample_result == []

    def test_sample_from_empty_range(self) -> None:
        """Test sampling from empty range."""
        empty_population = Container(int, [])

        sample_result = sample(empty_population, 5)
        assert sample_result == []
